plasmapy/classes/species.py

```diff
@@ -0,0 +1,231 @@
+        Charge and mass of particle.
+    eff_q: `astropy.units.Quantity`
+    eff_m: `astropy.units.Quantity`
+        Total charge and mass of macroparticle.
+    kinetic_energy
+    kinetic_energy_history
+
+    Examples
+    ----------
+    See `plasmapy/examples/particle-stepper.ipynb.`
+
+    """
+    def __init__(self, plasma, particle_type='p', n=1, scaling=1,
+                 dt=np.inf, nt=np.inf):
+        self.q = atomic.charge_state(particle_type) * constants.e.si
+        self.m = atomic.ion_mass(particle_type)
+        self.N = int(n)
+        self.scaling = scaling
+        self.eff_q = self.q * scaling
+        self.eff_m = self.m * scaling
+
+        self.plasma = plasma
+        if np.isinf(dt) and np.isinf(nt):
+            raise ValueError("Both dt and nt are infinite.")
+
+        self.dt = dt
+        assert self.dt.si.unit == u.s
+        self.NT = int(nt)
+        self.t = np.arange(nt) * dt
+
+        self.x = np.zeros((n, 3), dtype=float) * u.m
+        self.v = np.zeros((n, 3), dtype=float) * (u.m / u.s)
+        self.name = particle_type
+
+        self.position_history = np.zeros((self.NT, *self.x.shape),
+                                         dtype=float) * u.m
+        self.velocity_history = np.zeros((self.NT, *self.v.shape),
+                                         dtype=float) * (u.m / u.s)
+        self._B_interpolator = RegularGridInterpolator(
+            (self.plasma.x.si.value,
+             self.plasma.y.si.value,
+             self.plasma.z.si.value),
+            self.plasma.magnetic_field.T.si.value,
+            method="linear",
+            bounds_error=True)
+
+        self._E_interpolator = RegularGridInterpolator(
+            (self.plasma.x.si.value,
+             self.plasma.y.si.value,
+             self.plasma.z.si.value),
+            self.plasma.electric_field.T.si.value,
+            method="linear",
+            bounds_error=True)
+
+    def _interpolate_fields(self):
+        interpolated_b = self._B_interpolator(self.x.si.value) * u.T
+        interpolated_e = self._E_interpolator(self.x.si.value) * u.V / u.m
+        return interpolated_b, interpolated_e
+
+    @property
+    def kinetic_energy(self):
+        """
+        Calculates the kinetic energy for each macroparticle.
+
+        Returns
+        --------
+        `astropy.units.Quantity`
+            Array of kinetic energies, shape (n,).
+        """
+        return (self.v ** 2).sum(axis=-1) * self.eff_m / 2
+
+    @property
+    def kinetic_energy_history(self):
+        """
+        Calculates the kinetic energy history for each particle.
+
+        Returns
+        --------
+        `astropy.units.Quantity`
+            Array of kinetic energies, shape (nt, n).
+        """
+        return (self.velocity_history ** 2).sum(axis=-1) * self.eff_m / 2
+
+    def boris_push(self, init=False):
+        """
+        Implements the Boris algorithm for moving particles and updating their
+        velocities.
```

Comment:
It might be worth having a little more description, some maths and/or a link in the dosctring, for those who aren't familiar with the algorithm.